### PR TITLE
New datasources: kubernetes_namespace & kubernetes_all_namespaces

### PR DESCRIPTION
**Pull request by alexsomesan**
This change introduces a couple of new data sources for reading Kubernetes Namespace resources.
* `kubernetes_namespace` is a standard data source that exposes the attributes of a given namespace.
* `kubernetes_all_namespaces` is an enumerating data source that lists all available namespaces within a cluster. It can be used to test for the presence of a specific namespace or for applying a resource to all namespaces.

Test results:
```
» make testacc TESTARGS="-run '^TestAccKubernetes.*Namespace.*'"                                                             alex@MacBook-Pro-2
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test "./kubernetes" -v -run '^TestAccKubernetes.*Namespace.*' -timeout 120m
=== RUN   TestAccKubernetesDataSourceAllNamespaces_basic
--- PASS: TestAccKubernetesDataSourceAllNamespaces_basic (0.75s)
=== RUN   TestAccKubernetesDataSourceNamespace_basic
--- PASS: TestAccKubernetesDataSourceNamespace_basic (0.54s)
=== RUN   TestAccKubernetesNamespace_basic
--- PASS: TestAccKubernetesNamespace_basic (10.48s)
=== RUN   TestAccKubernetesNamespace_importBasic
--- PASS: TestAccKubernetesNamespace_importBasic (7.93s)
=== RUN   TestAccKubernetesNamespace_generatedName
--- PASS: TestAccKubernetesNamespace_generatedName (7.76s)
=== RUN   TestAccKubernetesNamespace_withSpecialCharacters
--- PASS: TestAccKubernetesNamespace_withSpecialCharacters (7.53s)
=== RUN   TestAccKubernetesNamespace_importGeneratedName
--- PASS: TestAccKubernetesNamespace_importGeneratedName (7.68s)
=== RUN   TestAccKubernetesNamespace_deleteTimeout
--- PASS: TestAccKubernetesNamespace_deleteTimeout (7.61s)
PASS
ok  	github.com/terraform-providers/terraform-provider-kubernetes/kubernetes	52.597s
------------------------------------------------------------
```

### CODE REVIEW

**Comment by alexsomesan** [comment]
Fixes #613

**Comment by pdecat** [comment]
Hi @alexsomesan, Nice!

But shouldn't singular datasources error on missing match?
See https://github.com/hashicorp/terraform/issues/16380#issuecomment-517282340

Also, no doc update? :wink:

**Comment by alexsomesan** [comment]
Yeah, good point about the doc. I'm juggling too many things at once. I'll push the docs ASAP.

As for the comment in Terraform core, I think there is a distinction to be made and I've added a clarification on issue #613 

I think that if the resource will eventually be managed by Terraform (e.g. asking for its prior state to condition  its eventual creation) the proper action should be to import it.

The datasource approach should only be used if the resource queried is entirely and permanently outside the scope of Terraform. Personal POV from being a TF user in the past.

**Comment by pdecat** [comment]
Agreed it is better to manage the resource with terraform.

Thing is this datasource may also be used to reference namespaces across modules.

Maybe making the datasource plural, and allow it to match zero, one or more namespaces on several criteria, e.g. name matches regex, would be more useful in the long term.

**Comment by alexsomesan** [comment]
I did initially consider the option of a simpler datasource that just lists all namespaces (they are global resources afterall). I went with this option to favor familiarity for users.

I am still able to add the extra listing datasource and revert this one to failing on 404. 
Would you find that to be a good user experience?

**Comment by pdecat** [comment]
I believe that would be more in line with what other providers do, e.g.:

* AWS:
  * `aws_availability_zone`: https://www.terraform.io/docs/providers/aws/d/availability_zone.html / https://github.com/terraform-providers/terraform-provider-aws/blob/v2.40.0/aws/data_source_aws_availability_zone.go#L75
  * `aws_availability_zone`: https://www.terraform.io/docs/providers/aws/d/availability_zones.html

* GCP:
  * `google_project`: https://www.terraform.io/docs/providers/google/d/google_project.html / https://github.com/terraform-providers/terraform-provider-google/blob/v2.20.0/google/data_source_google_project.go#L33 / https://github.com/terraform-providers/terraform-provider-google/blob/v2.20.0/google/resource_google_project.go#L294
  * `google_projects`: https://www.terraform.io/docs/providers/google/d/google_projects.html

File: website/docs/d/all_namespaces.html.markdown [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+page_title: "Kubernetes: kubernetes_all_namespaces"
+sidebar_current: "docs-kubernetes-data-source-all-namespaces"
+description: |-
+  Lists all namespaces within a cluster.
+---
+
+# kubernetes_all_namespaces
+
+This data source provides a mechanism for listing the names of all available namespaces in a Kubernetes cluster.
+It can be used check for existence of a specific namespaces or to apply another resource to all or a subset of existing namespaces in a cluster.
```
Comment (pdecat) [inline website/docs/d/all_namespaces.html.markdown]:
```suggestion
It can be used to check for existence of a specific namespaces or to apply another resource to all or a subset of existing namespaces in a cluster.
```

File: website/docs/d/all_namespaces.html.markdown [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+In Kubernetes, namespaces provide a scope for names and are intended as a way to divide cluster resources between multiple users.
```
Comment (pdecat) [inline website/docs/d/all_namespaces.html.markdown]: What about using a description closer to the official one which is a bit more explicit?

```suggestion
In Kubernetes, namespaces are intended for use in environments with many users spread across multiple teams, or projects. For clusters with a few to tens of users, you should not need to create or think about namespaces at all. Start using namespaces when you need the features they provide. 

Namespaces provide a scope for names. Names of resources need to be unique within a namespace, but not across namespaces. Namespaces can not be nested inside one another and each Kubernetes resource can only be in one namespace.

Namespaces are a way to divide cluster resources between multiple users (via resource quota).
```

From https://kubernetes.io/docs/concepts/overview/working-with-objects/namespaces/#when-to-use-multiple-namespaces

File: website/docs/d/all_namespaces.html.markdown [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+output "all-ns" {
+  value = data.kubernetes_all_namespaces.allns.namespaces
+}
+
+output "ns-present" {
+  value = contains(data.kubernetes_all_namespaces.allns.namespaces, "kube-system")
+}
+
+```
```
Comment (pdecat) [inline website/docs/d/all_namespaces.html.markdown]: Datasources usually have an attribute reference section:
````suggestion
```

## Attribute Reference

The following attributes are exported:

* `namespaces` - A list of existing namespace names.
````

See:
- https://github.com/terraform-providers/terraform-provider-kubernetes/blame/v1.10.0/website/docs/d/secret.html.markdown#L51-L54
- https://github.com/terraform-providers/terraform-provider-google/blame/v2.20.0/website/docs/d/datasource_compute_region_instance_group.html.markdown#L58-L69
- https://github.com/terraform-providers/terraform-provider-null/blame/v2.1.2/website/docs/data_source.html.markdown#L55-L64

File: kubernetes/data_source_kubernetes_all_namespaces.go [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+	nsRaw, err := conn.CoreV1().Namespaces().List(metav1.ListOptions{})
+	if err != nil {
+		log.Printf("[DEBUG] Received error: %#v", err)
+		return err
+	}
+	namespaces := make([]string, len(nsRaw.Items))
+	for i, v := range nsRaw.Items {
+		namespaces[i] = string(v.Name)
+	}
+	log.Printf("[INFO] Received namespaces: %#v", namespaces)
```
Comment (pdecat) [inline kubernetes/data_source_kubernetes_all_namespaces.go]: Probably `DEBUG` level stuff.

File: kubernetes/data_source_kubernetes_namespace.go [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+	d.SetId(name)
+	conn := meta.(*KubeClientsets).MainClientset
+
+	log.Printf("[INFO] Reading namespace %s", name)
+	namespace, err := conn.CoreV1().Namespaces().Get(name, meta_v1.GetOptions{})
+	if err != nil {
+		log.Printf("[DEBUG] Received error: %#v", err)
+		return err
+	}
+	log.Printf("[INFO] Received namespace: %#v", namespace)
```
Comment (pdecat) [inline kubernetes/data_source_kubernetes_namespace.go]: Probably `DEBUG` level stuff.

File: website/docs/d/namespace.html.markdown [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+
+## Nested Blocks
+
+### `metadata`
+
+#### Arguments
+
+* `name` - (Required) Name of the namespace, must be unique. For more info see [Kubernetes reference](http://kubernetes.io/docs/user-guide/identifiers#names)
+
+#### Attributes
```
Comment (pdecat) [inline website/docs/d/namespace.html.markdown]: Shouldn't this go in an attribute reference section?

```suggestion
## Attribute Reference

The following attributes are exported:
```

File: website/docs/d/namespace.html.markdown [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+
+* `annotations` - (Optional) An unstructured key value map stored with the namespace that may be used to store arbitrary metadata.
```
Comment (pdecat) [inline website/docs/d/namespace.html.markdown]: Probably not needed to specify `(Optional)` as it is an export only attribute, is it?

File: kubernetes/data_source_kubernetes_all_namespaces.go [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+	}
+	idsum := sha256.New()
+	for _, v := range namespaces {
+		_, err := idsum.Write([]byte(v))
+		if err != nil {
+			return err
+		}
+	}
+	id := fmt.Sprintf("%x", idsum.Sum(nil))
+	d.SetId(id)
```
[the remark this answers is not in the document]
Comment (alexsomesan) [reply kubernetes/data_source_kubernetes_all_namespaces.go]: No guidelines that I know of. I just went for this approach so that no changes are "registered" if the data (list of namespaces) didn't change.

I can see the motivations behind the Google (save an API call if the query filter didn't change) and the AWS one (ensure a fresh query is always done). I guess the AWS approach would have worked for this case as well.

File: kubernetes/provider.go [hunk 9]
```diff
@@ -135,9 +135,11 @@ func Provider() terraform.ResourceProvider {
 		},
 
 		DataSourcesMap: map[string]*schema.Resource{
-			"kubernetes_secret":        dataSourceKubernetesSecret(),
-			"kubernetes_service":       dataSourceKubernetesService(),
-			"kubernetes_storage_class": dataSourceKubernetesStorageClass(),
+			"kubernetes_all_namespaces": dataSourceKubernetesAllNamespaces(),
```
[the remark this answers is not in the document]
Comment (alexsomesan) [reply kubernetes/provider.go]: I did contemplate that possibility of future filters, so that's a valid point you're making.
Do you see filtering based on labels / annotations as something useful to have?
Can't think of anything else you'd want to filter on.

File: website/docs/d/namespace.html.markdown (hunk 7 above)
Comment (alexsomesan) [reply website/docs/d/namespace.html.markdown]: true!

File: kubernetes/data_source_kubernetes_namespace.go (hunk 5 above)
Comment (alexsomesan) [reply kubernetes/data_source_kubernetes_namespace.go]: Tanken from the resource read. It's INFO over there too, also in every other resource. Thought I'd stay consistent.

File: kubernetes/data_source_kubernetes_all_namespaces.go (hunk 4 above)
Comment (alexsomesan) [reply kubernetes/data_source_kubernetes_all_namespaces.go]: It's INFO in all other reads.

File: kubernetes/provider.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (pdecat) [reply kubernetes/provider.go]: Main use case I imagined was to filter on the `metadata.0.name` value, e.g. applying things like default network policies to namespaces matching `.*-public` or `.*-private`, but labels and annotations are good candidates too.

File: kubernetes/data_source_kubernetes_all_namespaces.go (hunk 4 above)
Comment (pdecat) [reply kubernetes/data_source_kubernetes_all_namespaces.go]: Indeed, my bad.

**Comment by carlpett** [comment]
Just a note from the sidelines here, that while I think it is great with a datasource for namespaces, as the one filing #613, I do not agree that it actually "fixes" that issue as I stated it. It does solve some tangential issues others who chimed in in the thread had, but those where to my mind at least different.
To reiterate what I was after, it was more about dealing with race conditions where some tools (notably Helm 2.x) will create namespaces that do not already exist. We prefer to manage namespaces with Terraform, but due to those tools racing with Terraform once we create our clusters, we will *sometimes* have failures during our Terraform runs which create new clusters.

Again, this is not to throw any shade on this PR, I think it is great and very needed! :tada: I just don't think it is that closely related to the issue.

**Comment by Nightbr** [comment]
Hey, any blockers here in order to merge this PR? I would really love to see this feature & can help if necessary.

File: README.md [hunk 10]
```diff
@@ -12,7 +12,7 @@
 
 
 -	[Terraform](https://www.terraform.io/downloads.html) 0.10.x
--	[Go](https://golang.org/doc/install) 1.11.x (to build the provider plugin)
+-	[Go](https://golang.org/doc/install) 1.14.x (to build the provider plugin)
```
Comment (dak1n1) [inline README.md:15]: @jrhouston I noticed the Go version hadn't been updated in a while... is this ok to do? Or should I make it a separate PR? I updated it as part of some troubleshooting, but it's not necessarily needed for this PR.

Comment (jrhouston) [reply README.md:15]: This is fine!
